chains/evm/contracts/ContractFactory.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.24;
+
+import {ITypeAndVersion} from "@chainlink/contracts/src/v0.8/shared/interfaces/ITypeAndVersion.sol";
+
+import {Ownable2StepMsgSender} from "@chainlink/contracts/src/v0.8/shared/access/Ownable2StepMsgSender.sol";
+
+import {Create2} from "@openzeppelin/contracts@5.0.2/utils/Create2.sol";
+import {EnumerableSet} from "@openzeppelin/contracts@5.0.2/utils/structs/EnumerableSet.sol";
+
+/// @notice A contract for deploying and configuring contracts via CREATE2.
+/// @dev This contract is used to deploy static, user-facing contracts (e.g. proxies)
+/// such that the resulting addresses are the same across multiple chains. Optionally, you can provide
+/// one or more initialization calls (i.e. transfer ownership, configure roles) to perform post-deployment.
+/// To achieve deterministic addresses across chains, this contract must be deployed with a reserved key.
+/// This is because the factory address is used in the CREATE2 address computation.
+contract ContractFactory is ITypeAndVersion, Ownable2StepMsgSender {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  error CallFailed(uint256 index, bytes result);
+  error CallerNotAllowed(address caller);
+
+  event CallerAdded(address indexed caller);
+  event CallerRemoved(address indexed caller);
+  event ContractDeployed(address indexed contractAddress);
+
+  string public constant override typeAndVersion = "ContractFactory 1.7.0";
+
+  /// @notice Addresses that are allowed to call createAndCall.
+  EnumerableSet.AddressSet private s_allowList;
+
+  constructor(
```

Comment:
If we don't do this, someone could snipe our address? We could also include the sender in the hash instead, right? So only our key could derive the same address. That would mean we don't have any config or state in this contract